src/VecSim/algorithms/hnsw/hnsw_tiered.h

```diff
@@ -73,3 +108,90 @@ class TieredHNSWIndex : public VecSimTieredIndex<DataType, DistType> {
         return this->index->setLastSearchMode(mode);
     }
 };
+
+/**
+ ******************************* Implementation **************************
+ */
+
+/* Helper methods */
+template <typename DataType, typename DistType>
+void TieredHNSWIndex<DataType, DistType>::executeInsertJobWrapper(void *job) {
+    auto *insert_job = reinterpret_cast<HNSWInsertJob *>(job);
+    reinterpret_cast<TieredHNSWIndex<DataType, DistType> *>(insert_job->index)
+        ->executeInsertJob(insert_job);
+}
+
+template <typename DataType, typename DistType>
+void TieredHNSWIndex<DataType, DistType>::submitSingleJob(AsyncJob *job) {
+    auto **jobs = array_new<AsyncJob *>(1);
+    jobs = array_append(jobs, job);
+    this->SubmitJobsToQueue(this->jobQueue, (void **)jobs, 1);
+    array_free(jobs);
+}
+
+/******************** Index API ****************************************/
+
+template <typename DataType, typename DistType>
+TieredHNSWIndex<DataType, DistType>::TieredHNSWIndex(HNSWIndex<DataType, DistType> *hnsw_index,
+                                                     TieredIndexParams tieredParams)
+    : VecSimTieredIndex<DataType, DistType>(hnsw_index, tieredParams),
+      labelToInsertJobs(this->allocator), idToRepairJobs(this->allocator),
+      idToSwapJob(this->allocator) {
+    this->UpdateIndexMemory(this->memoryCtx, this->getAllocationSize());
+}
+
+template <typename DataType, typename DistType>
+TieredHNSWIndex<DataType, DistType>::~TieredHNSWIndex() {
+    // Delete all the pending insert jobs.
+    for (auto jobs : this->labelToInsertJobs) {
+        for (auto *job : jobs.second) {
+            delete job;
+        }
+    }
+}
+
+template <typename DataType, typename DistType>
+size_t TieredHNSWIndex<DataType, DistType>::indexSize() const {
+    return this->index->indexSize() + this->flatBuffer->indexSize();
+}
+
+template <typename DataType, typename DistType>
+size_t TieredHNSWIndex<DataType, DistType>::indexCapacity() const {
+    return this->index->indexCapacity() + this->flatBuffer->indexCapacity();
+}
+
+template <typename DataType, typename DistType>
+size_t TieredHNSWIndex<DataType, DistType>::indexLabelCount() const {
+    return this->flatBuffer->indexLabelCount() + this->index->indexLabelCount();
+}
+
+template <typename DataType, typename DistType>
+int TieredHNSWIndex<DataType, DistType>::addVector(const void *blob, labelType label,
+                                                   bool overwrite_allowed) {
+    /* Note: this currently doesn't support overriding (assuming that the label doesn't exist)! */
+    this->flatIndexGuard.lock();
+    if (this->flatBuffer->indexCapacity() == this->flatBuffer->indexSize()) {
+        this->flatBuffer->increaseCapacity();
+    }
```

Comment:
Why is this needed? isn't that part of `addVector`?

Reply:
No, we intentionally moved the `resize` outside the `addVector`, as the `resize` requires a write lock in HNSW, while add vector is done under read lock. Hence, the generic API now has a separate `increaseCapacity` call (which is used in flat index as well)